server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java

```diff
@@ -53,64 +113,198 @@
         }
+
+        return createAnnotationResponse(annotation);
+    }
+
+    @Override
+    public AnnotationResponse updateAnnotationVisibility(UpdateAnnotationVisibilityCmd cmd) {
+        String uuid = cmd.getUuid();
+        Boolean adminsOnly = cmd.getAdminsOnly();
         AnnotationVO annotation = annotationDao.findByUuid(uuid);
-        annotationDao.remove(annotation.getId());
+        if (annotation != null && isCallingUserAdmin()) {
+            if(LOGGER.isDebugEnabled()) {
+                LOGGER.debug("updating annotation visibility: " + uuid);
+            }
+            annotation.setAdminsOnly(adminsOnly);
+            annotationDao.update(annotation.getId(), annotation);
+        } else {
+            throw new CloudRuntimeException("Cannot update visibility for annotation: " + uuid);
+        }
         return createAnnotationResponse(annotation);
     }
 
+    private boolean isCallingUserAllowedToRemoveAnnotation(AnnotationVO annotation) {
+        if (annotation == null) {
+            return false;
+        }
+        if (isCallingUserAdmin()) {
+            return true;
+        }
+        UserVO callingUser = getCallingUserFromContext();
+        String annotationOwnerUuid = annotation.getUserUuid();
+        return annotationOwnerUuid != null && annotationOwnerUuid.equals(callingUser.getUuid());
+    }
+
+    private UserVO getCallingUserFromContext() {
+        CallContext ctx = CallContext.current();
+        long userId = ctx.getCallingUserId();
+        UserVO userVO = userDao.findById(userId);
+        if (userVO == null) {
+            throw new CloudRuntimeException("Cannot find a user with ID " + userId);
+        }
+        return userVO;
+    }
+
+    private boolean isCallingUserAdmin() {
+        UserVO userVO = getCallingUserFromContext();
+        long accountId = userVO.getAccountId();
+        AccountVO accountVO = accountDao.findById(accountId);
+        if (accountVO == null) {
+            throw new CloudRuntimeException("Cannot find account with ID + " + accountId);
+        }
+        Long roleId = accountVO.getRoleId();
+        Role role = roleService.findRole(roleId);
+        if (role == null) {
+            throw new CloudRuntimeException("Cannot find role with ID " + roleId);
+        }
+        return adminRoles.contains(role.getRoleType());
+    }
+
     private List<AnnotationVO> getAnnotationsForApiCmd(ListAnnotationsCmd cmd) {
         List<AnnotationVO> annotations;
-        if(cmd.getUuid() != null) {
+        String userUuid = cmd.getUserUuid();
+        String entityUuid = cmd.getEntityUuid();
+        String entityType = cmd.getEntityType();
+        String annotationFilter = isNotBlank(cmd.getAnnotationFilter()) ? cmd.getAnnotationFilter() : "all";
+        boolean isCallerAdmin = isCallingUserAdmin();
+        if ((isBlank(entityUuid) || isBlank(entityType)) && !isCallerAdmin && annotationFilter.equalsIgnoreCase("all")) {
```

Comment:
`StringUtils` has the method `isAnyEmpty`, where we can validate several strings at the same time.